### DISCUSSION TITLE
Validate from an array of possible literal values

### DISCUSSION THREAD

**Discussion by MikeRippon** [colinhacks/zod#500] [Ideas]
Hi, this can be done with `refine()` and `transform()`, but I'm just wondering if there is a shorter syntax for validating based on an array of literal values. It seems like it should be a reasonably common use case? Chaining `or` isn't really suitable if the array of values is declared and used elsewhere in the codebase.

My current situation is based on enums, but I don't see why this would have to be enum-specific, and could potentially work with any value:

```typescript
enum Fruit {
    Strawberry,
    Apple,
    Kiwi,
    Orange,
    Lime,
    Lemon,
};

// Used in various places, e.g. to drive dropdown values
const citrusFruits = [Fruit.Lime, Fruit.Lemon, Fruit.Orange] as const;
```

Would be nice:

```typescript
// Could probably infer type, provided the array is declared `as const`?
const citrusFruitSchema = z.oneOf(citrusFruits);
```

Current workaround

```typescript
type CitrusFruit = typeof citrusFruits[number];
const citrusFruitSchema = z.any()
    .refine(
        fruit => (citrusFruits as readonly Fruit[]).includes(fruit),
        `Fruit must be one of [${citrusFruits.join(', ')}]`
    )
    .transform(fruit => fruit as CitrusFruit);
```

Sorry if I've missed something in the docs!

**Comment by StefanTerdell** [comment]
As long as the enum represents a string value, z.enum() should do the trick. Does that help?

```typescript
enum Fruit {
    Strawberry = "Strawberry",
    Apple = "Apple",
    Kiwi = "Kiwi",
    Orange = "Orange",
    Lime = "Lime",
    Lemon = "Lemon",
};

const citrusFruits = [Fruit.Lime, Fruit.Lemon, Fruit.Orange] as const;

const citrusFruitsSchema = z.enum(citrusFruits)
```

**Reply by MikeRippon** [reply]
That absolutely does thank you!

**Reply by MikeRippon** [reply]
Note: this works for validation, but it only infers the generic `Fruit` type
